src/software/simulated_tests/non_terminating_validation_functions/ball_always_moves_forward_validation.cpp

```diff
@@ -0,0 +1,14 @@
+#include "software/simulated_tests/non_terminating_validation_functions/ball_always_moves_forward_validation.h"
+
+void ballAlwaysMovesForward(std::shared_ptr<World> world_ptr,
+                            ValidationCoroutine::push_type& yield)
+{
+    static double previous_x_pos = world_ptr->field().friendlyGoalCenter().x();
+
```

Comment:
I see you assign `previous_x_pos = current_x_pos` at the end of the function; I may be missing something here, but how does the validation function know to skip this initial assignment when it repeats itself?

Reply:
This is quirky C++ thing but there will always only be one copy of this variable in the function when you declare it static and the variable will retain its value when you keep running the function. https://stackoverflow.com/a/6223689

Reply:
Whoops I must've missed this. We should be keeping state using coroutines, i.e.
```
double previous_x_pos = world_ptr->field().friendlyGoalCenter().x();
while(world_ptr->ball().position().x() >= previous_x_pos)
{
    previous_x_pos = world_ptr->ball().position().x();
    yield();
}
FAIL() << "Ball has moved backward";

```

Reply:
updated